### PR TITLE
NF: remove useless escape in regexp

### PR DESCRIPTION
**Pull request by Arthur-Milchior**

### CODE REVIEW

**Comment by Arthur-Milchior** [comment]
> > Task :AnkiDroid:connectedDebugAndroidTest
> Starting 0 tests on test(AVD) - 4.1.2
> Tests on test(AVD) - 4.1.2 failed: Instrumentation run failed due to 'java.util.regex.PatternSyntaxException'
> com.android.build.gradle.internal.testing.ConnectedDevice > No tests found.[test(AVD) - 4.1.2] FAILED 
> No tests found. This usually means that your test classes are not in the form that your test runner expects (e.g. don't inherit from TestCase or lack @Test annotations).
> > Task :AnkiDroid:connectedDebugAndroidTest FAILED
> FAILURE: Build failed with an exception.
> * What went wrong:
> Execution failed for task ':AnkiDroid:connectedDebugAndroidTest'.


Can you please help me understand ? In particular, I don't see what AnkiDroid:connectedDebugAndroidTest actually is

**Comment by mikehardy** [comment]
connected  == it's running on an emulator
Debug == the build type (alternative is Release)
AndroidTest == running a test, specifically the batch from the androidTest directory

This happens sometimes but is usually just a transient failure, I'll try restarting it

**Comment by mikehardy** [comment]
Never mind, I just looked at one of the bulids and it's a total failure on all APIs for what looks like a real reason

**Comment by Arthur-Milchior** [comment]
Oh, it's clearly a real reason. It's clearly related to this very PR, because it suddenly got a `PatternSyntaxException` which is exactly what I would expect if there is any error related to simplifying regexp by removing escape.

What makes no sens to me at all is: how can change in the source code have such a big impact that no tests are detected and nothing gets executed. If there were any change related to travis rules. There is not even a single change in a test file or in a non java file.
If at least the build was failing, I could see where it fails. But this problem is entirely new to me

**Comment by david-allison-1** [comment]
> Oh, it's clearly a real reason. It's clearly related to this very PR, because it suddenly got a `PatternSyntaxException` which is exactly what I would expect if there is any error related to simplifying regexp by removing escape.
> 
> What makes no sens to me at all is: how can change in the source code have such a big impact that no tests are detected and nothing gets executed. If there were any change related to travis rules. There is not even a single change in a test file or in a non java file.
> If at least the build was failing, I could see where it fails. But this problem is entirely new to me

Probably a crash in the static constructor of a class

**Comment by Arthur-Milchior** [comment]
Dichotomy led to find the troublesome one. I added a comment to it. If anyone have an explanation, I'd love it

**Comment by david-allison-1** [comment]
Could you confirm these lines are covered by appropriate tests - if the auto-fix is broken, it can also be broken in ways that still compile

**Comment by Arthur-Milchior** [comment]
No, I don't know the test base well enough to know whether all changes are covered.

I made some test. If I uses the exact same pattern in a class which contains just a main and which execute a match with those patterns, then it works exactly as intended.

If I inline those members instead of having them static, the bug disappear too. The bug seems to appear only when this compilation occurs in a static variable and disappear when the compilation of the regexp occurs in the method

**Comment by Arthur-Milchior** [comment]
I had my answer. Android does not uses the same library than Java. https://stackoverflow.com/questions/45074813/regex-pattern-error-in-android-when-matching-closing-brace
So I added back escape for }

I'll try to make bug report for IntelliJ rewritting rule and if I can figure out how to do it, for android manual to indicate the difference with java. Seems pretty important

**Comment by david-allison-1** [comment]
> I had my answer. Android does not uses the same library than Java. https://stackoverflow.com/questions/45074813/regex-pattern-error-in-android-when-matching-closing-brace
> So I added back escape for }
> 
> I'll try to make bug report for IntelliJ rewritting rule and if I can figure out how to do it, for android manual to indicate the difference with java. Seems pretty important

Doesn't this mean that any of the other changes that fix `}` are incorrect

**Comment by Arthur-Milchior** [comment]
Yeah. I removed them (but i see the force push failed)

**Review by david-allison-1 — APPROVED** [review]

**Comment by mikehardy** [comment]
Still failing Travis :thinking:

**Comment by david-allison-1** [comment]
```
> Task :AnkiDroid:compileReleaseJavaWithJavac
/Users/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java:243: error: <identifier> expected
            t.put(type, t.getString(type).replaceAll("\\{\\{cloze:1:(.+?)\\}\\}", "{{cloze:$1}}"));
                 ^
/Users/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java:243: error: not a statement
            t.put(type, t.getString(type).replaceAll("\\{\\{cloze:1:(.+?)\\}\\}", "{{cloze:$1}}"));
                  ^
/Users/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java:243: error: ';' expected
            t.put(type, t.getString(type).replaceAll("\\{\\{cloze:1:(.+?)\\}\\}", "{{cloze:$1}}"));
                      ^
/Users/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java:243: error: ';' expected
            t.put(type, t.getString(type).replaceAll("\\{\\{cloze:1:(.+?)\\}\\}", "{{cloze:$1}}"));
                                                                                                 ^
4 errors
```

File: AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java [hunk 1]
```diff
@@ -239,6 +239,7 @@ private static void _upgradeClozeModel(Collection col, Model m) throws ConfirmMo
         // convert first template
         JSONObject t = m.getJSONArray("tmpls").getJSONObject(0);
         for (String type : new String[] { "qfmt", "afmt" }) {
+            @SuppressWarnings("RegExpRedundantEscape") // In Android, } should be escaped
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java:242]: This needs to be a comment, not an annotation

**Pull request merged**